File: ledger/Assets/Scripts/Core/VoiceBank.cs
using System;
using System.Collections.Generic;

namespace Ledger.Core
{
    /// WHICH RECORDING IS THIS LINE — audit items 5 and 7, which turn out to
    /// be the same question.
    ///
    /// Item 7 said there was no voice bus. There is one now, and it needs to
    /// be handed a filename. Item 5 said nothing specified determinism: a
    /// cloner given the same reference and the same text can produce a
    /// different take, so if the bank is ever regenerated every character
    /// drifts. Both are answered by one rule:
    ///
    ///   **A clip is named by a hash of who is speaking and what they say,
    ///   and the generator seeds itself from the same hash.**
    ///
    /// So the game can ask for a line by computing the name, the generator
    /// can produce it by computing the same name, neither has to consult a
    /// manifest, and regenerating the bank next month yields byte-identical
    /// takes rather than nineteen people who now sound slightly different.
    ///
    /// THE HASH IS WRITTEN OUT BY HAND ON PURPOSE. `string.GetHashCode()` is
    /// randomised per process in modern .NET; it is stable in Unity's runtime
    /// today and that is an accident of the runtime, not a promise. A cache
    /// key that changes when the engine updates would silently orphan the
    /// entire bank, and the failure would look like every voice vanishing at
    /// once for no reason anybody could trace. FNV-1a is nine lines and it is
    /// the same number on every platform, in every process, forever.
    public static class VoiceBank
    {
        /// The background voices the crowd draws from — and these are not a
        /// number I picked, they are the six crowd entries the casting sheet
        /// actually funds (`tools/voice-fetch`, tier "crowd").
        ///
        /// SIX IS THIN and it is written down here rather than papered over.
        /// The ear picks out a repeated voice on a busy street inside a
        /// minute, and a city where every seventh person shares a throat
        /// reads worse than one with no voices at all. The fix is more crowd
        /// entries in the casting sheet, which costs generation time — not a
        /// larger constant here, which would only name files nobody is ever
        /// going to produce and turn a thin bank into a silent one.
        public static readonly string[] PoolMasculine = { "crowd_m1", "crowd_m2", "crowd_m3" };
        public static readonly string[] PoolFeminine = { "crowd_f1", "crowd_f2", "crowd_f3" };

        public static int PoolVoices => PoolMasculine.Length + PoolFeminine.Length;

        /// The named voices, copied from the casting sheet the fetcher works
        /// from (`tools/voice-fetch`, tiers "principal" and "street").
        ///
        /// AND THESE DO NOT ALL MATCH THE ROSTER. The game's gossipers use
        /// ids like `sera`, `aldous`, `danny`, `halvard`, `june`, `zlata`;
        /// the casting sheet has `kest`, `vesna`, `marla`. Some are the same
        /// person under two names and some are people nobody has cast. The
        /// consequence is contained and deliberate — an id not in this set
        /// draws a crowd voice rather than throwing — but it means a named
        /// character can quietly end up sounding like a passer-by, so the
        /// reconciliation is a casting task somebody has to actually do.
        public static readonly HashSet<string> Cast = new HashSet<string>
        {
            "lena", "rocco", "ellis", "reese", "kest",
            "sam", "ada", "vesna", "marla", "joey", "rita", "hal", "emil",
        };

        /// The byte between the voice and the words when they are hashed.
        ///
        /// NOT AN EMPTY STRING. Concatenating with nothing between them makes
        /// ("ab","cd") and ("abc","d") hash identically — harmless for the
        /// filename, where the voice is also the folder, and not harmless for
        /// `Seed`, where two different lines would then generate as the same
        /// take. A unit separator, because it cannot occur in a line of
        /// dialogue, written as an escape because a raw control byte sitting
        /// in a source file is a thing nobody should have to discover.
        const string Sep = "\u001f";

        /// FNV-1a, 32-bit. Stable across platforms, processes and runtimes.
        public static uint Hash(string s)
        {
            unchecked
            {
                uint h = 2166136261u;
                if (s == null) return h;
                for (int i = 0; i < s.Length; i++)
                {
                    h ^= s[i];
                    h *= 16777619u;
                }
                return h;
            }
        }

        /// What the hash is actually taken over.
        ///
        /// Whitespace is collapsed because a line that gained a double space
        /// in an edit is the same line and must not orphan its recording.
        /// CASE IS KEPT: capitals change how a text-to-speech engine reads a
        /// sentence — emphasis, initialisms, the difference between "no" and
        /// "NO" — so two lines differing only in case are two different
        /// performances and deserve two different files.
        public static string Normalise(string text)
        {
            if (string.IsNullOrEmpty(text)) return "";
            var sb = new System.Text.StringBuilder(text.Length);
            bool space = false;
            foreach (var ch in text)
            {
                if (char.IsWhiteSpace(ch)) { space = sb.Length > 0; continue; }
                if (space) { sb.Append(' '); space = false; }
                sb.Append(ch);
            }
            return sb.ToString();
        }

        /// The clip's name: `voice/hash`, and nothing else.
        ///
        /// No slot id, no line index, no scene, no act. Every one of those is
        /// a thing that can be renamed while the words stay identical, and a
        /// filename that changes when the words did not is how a bank rots.
        /// Returns null for anything unspeakable, so a caller cannot get a
        /// plausible-looking path for a line that does not exist.
        public static string ClipName(string voiceId, string text)
        {
            var t = Normalise(text);
            if (string.IsNullOrEmpty(voiceId) || t.Length == 0) return null;
            return voiceId + "/" + Hash(voiceId + Sep + t).ToString("x8");
        }

        /// The generator's random seed for that clip.
        ///
        /// Same inputs, same seed, same take — which is the whole of audit
        /// item 5. Non-negative because every RNG this project uses takes an
        /// `int` and half of them treat a negative as an error.
        public static int Seed(string voiceId, string text)
        {
            var t = Normalise(text);
            if (string.IsNullOrEmpty(voiceId) || t.Length == 0) return 0;
            return (int)(Hash(voiceId + Sep + t) & 0x7fffffff);
        }

        /// Which voice a given speaker uses.
        ///
        /// A cast member IS their voice — Rocco is cast, recorded and named.
        /// Everybody else draws from the crowd pool, deterministically by id,
        /// so the same walker sounds like the same person every time you pass
        /// them and nobody has to store which voice they got.
        ///
        /// `masculine` is null when the caller does not know, and today every
        /// caller is null: `SpeechBubble` gets a Transform, not a resident
        /// record, so the speaker's gender is not in reach at the point the
        /// line is spoken. Said out loud rather than defaulted quietly —
        /// unknown means the pool is drawn from as a whole, which will
        /// sometimes put a woman's voice on a man until the population record
        /// is threaded through to the call site.
        public static string VoiceFor(string speakerId, ICollection<string> castVoiceIds,
                                      bool? masculine = null)
        {
            if (string.IsNullOrEmpty(speakerId)) return null;
            if (castVoiceIds != null && castVoiceIds.Contains(speakerId)) return speakerId;
            uint h = Hash(speakerId);
            if (masculine == true) return PoolMasculine[h % (uint)PoolMasculine.Length];
            if (masculine == false) return PoolFeminine[h % (uint)PoolFeminine.Length];
            return (h % 2 == 0 ? PoolMasculine : PoolFeminine)
                   [(h / 2) % (uint)PoolMasculine.Length];
        }

        /// A stable seed for anything else that needs to vary by line and
        /// must not vary by run — the word-dropping in `Acoustics.AsHeard`
        /// being the one that already existed and was reaching for
        /// `string.GetHashCode()` to get it.
        public static int SeedFor(string text) => (int)(Hash(Normalise(text)) & 0x7fffffff);
    }
}

File: tools/voice-fetch/voice_bank.py
#!/usr/bin/env python3
"""The clip naming rule, in Python, byte-for-byte the same as the game's.

    python3 voice_bank.py --selftest

WHY THIS FILE IS A DUPLICATE ON PURPOSE.

`Assets/Scripts/Core/VoiceBank.cs` decides what a clip is called and what
seed generates it. The game computes that name to find a recording; the
generator has to compute the SAME name to write one. They are different
languages in different processes and there is no shared runtime between
them, so one of three things had to be true:

  a) the generator writes a manifest and the game reads it — one more file
     to keep in sync, and a game that cannot find a line without it;
  b) the game shells out to Python, which it obviously cannot;
  c) the rule is small enough to state twice and PINNED BY SHARED TEST
     VECTORS so a drift is caught rather than discovered.

(c), and the vectors below are asserted in CoreTests too. If either side
changes the hash, both sides go red on the same numbers rather than the bank
silently orphaning itself — which is the failure that would look like every
voice in the game vanishing at once for no traceable reason.

Keep this file and VoiceBank.cs in agreement. The test vectors are the
contract; the implementations are just two readings of it.
"""
import struct
import sys

# The six crowd voices the casting sheet actually funds. Not a number
# somebody liked -- see VoiceBank.cs on why six is thin and why the fix is
# casting rather than a larger constant here.
POOL_MASCULINE = ("crowd_m1", "crowd_m2", "crowd_m3")
POOL_FEMININE = ("crowd_f1", "crowd_f2", "crowd_f3")

CAST = frozenset({
    "lena", "rocco", "ellis", "reese", "kest",
    "sam", "ada", "vesna", "marla", "joey", "rita", "hal", "emil",
})

# Between the voice and the words. Not an empty string: concatenating with
# nothing between them makes ("ab","cd") and ("abc","d") hash identically.
SEP = "\u001f"


def hash32(s):
    """FNV-1a, 32-bit. The same number on every platform, in every process.

    Python's own hash() is randomised per process by default and C#'s
    string.GetHashCode() is randomised in modern .NET -- either would make a
    cache key that changes for reasons nobody can see."""
    h = 2166136261
    if s is None:
        return h
    # UTF-16 CODE UNITS, not Python characters. C# iterates `s[i]` over a
    # UTF-16 string, so a character outside the basic plane is TWO units
    # there and ONE here -- the two implementations would agree on every
    # line of dialogue anybody is likely to write and disagree on the first
    # emoji or rare glyph, which is the worst possible place to differ
    # because it would look like a one-off corrupt file.
    raw = s.encode("utf-16-le")
    for unit in struct.unpack("<%dH" % (len(raw) // 2), raw):
        h ^= unit
        h = (h * 16777619) & 0xFFFFFFFF
    return h


def normalise(text):
    """Collapse whitespace, keep case.

    Case is kept because capitals change how a text-to-speech engine reads a
    sentence, so "no" and "NO" are two performances and deserve two files."""
    if not text:
        return ""
    return " ".join(text.split())


def clip_name(voice_id, text):
    t = normalise(text)
    if not voice_id or not t:
        return None
    return "%s/%08x" % (voice_id, hash32(voice_id + SEP + t))


def seed(voice_id, text):
    t = normalise(text)
    if not voice_id or not t:
        return 0
    return hash32(voice_id + SEP + t) & 0x7FFFFFFF


def voice_for(speaker_id, cast=CAST, masculine=None):
    if not speaker_id:
        return None
    if cast and speaker_id in cast:
        return speaker_id
    h = hash32(speaker_id)
    if masculine is True:
        return POOL_MASCULINE[h % len(POOL_MASCULINE)]
    if masculine is False:
        return POOL_FEMININE[h % len(POOL_FEMININE)]
    pool = POOL_MASCULINE if h % 2 == 0 else POOL_FEMININE
    return pool[(h // 2) % len(pool)]


# THE CONTRACT, and it is written out rather than computed. Every one of
# these is asserted in CoreTests as well, against the C# implementation, so a
# drift on either side goes red on both with the same numbers instead of the
# bank silently orphaning itself.
#
# The astral entry is deliberate: it is the case where a naive Python port
# disagrees with C# about what a character is, and it disagrees SILENTLY.
VECTORS = [
    # (voice, text, clip name, seed)
    ("rocco", "He was at the yard on Tuesday.", "rocco/df92fd5e", 1603468638),
    ("lena", "He was at the yard on Tuesday.", "lena/1d5782f8", 492274424),
    ("crowd_m1", "Evening.", "crowd_m1/953df5cc", 356382156),
    # Outside the basic plane: two UTF-16 units in C#, one character in
    # Python. A naive port passes every other vector here and fails only
    # this one, which is why it is here.
    ("rocco", "Told you \U0001F600 nothing.", "rocco/f278f6c6", 1920530118),
]


def selftest():
    ok = fail = 0

    def check(cond, what, detail=""):
        nonlocal ok, fail
        if cond:
            ok += 1
        else:
            fail += 1
            print("  FAILED: %s %s" % (what, detail))

    check(hash32("") == 2166136261, "FNV-1a offset basis")
    check(hash32("a") == 0xe40c292c, "FNV-1a of 'a'", hex(hash32("a")))
    check(hash32("foobar") == 0xbf9cf968, "FNV-1a of 'foobar'", hex(hash32("foobar")))

    check(normalise("  two   words \n") == "two words", "whitespace collapses")
    check(normalise("No") != normalise("NO"), "case is kept")

    said = "He was at the yard on Tuesday."
    check(clip_name("rocco", said) == clip_name("rocco", "  He was at the yard on Tuesday.  "),
          "a re-indented line is the same recording")
    check(clip_name("rocco", said) != clip_name("lena", said),
          "two people saying the same words are two recordings")
    check(clip_name("rocco", said).startswith("rocco/"), "the voice is the folder")
    check(len(clip_name("rocco", said)) == len("rocco/") + 8, "voice plus eight hex digits")
    check(clip_name(None, said) is None and clip_name("rocco", "  ") is None,
          "an unspeakable line gets no plausible path")

    check(seed("rocco", said) == seed("rocco", said), "the same line seeds the same take")
    check(seed("rocco", said) >= 0, "the seed is non-negative")

    check(voice_for("rocco") == "rocco", "a cast member IS their voice")
    check(voice_for("resident_8817").startswith("crowd_"), "everybody else draws a crowd voice")
    check(voice_for("resident_8817", masculine=True).startswith("crowd_m"),
          "and a known gender gets a matching voice")
    reached = {voice_for("resident_%d" % i) for i in range(4000)}
    check(len(reached) == len(POOL_MASCULINE) + len(POOL_FEMININE),
          "four thousand walkers reach every crowd voice", str(sorted(reached)))

    # THE CROSS-LANGUAGE CONTRACT, asserted rather than printed. Printing
    # them for a human to compare by eye is how two implementations stay
    # "obviously the same" right up until they are not.
    for voice, text, want_name, want_seed in VECTORS:
        check(clip_name(voice, text) == want_name,
              "vector %s/%r names %s" % (voice, text[:24], want_name),
              str(clip_name(voice, text)))
        check(seed(voice, text) == want_seed,
              "vector %s/%r seeds %d" % (voice, text[:24], want_seed),
              str(seed(voice, text)))

    print("\n%d passed, %d failed" % (ok, fail))
    return 1 if fail else 0


if __name__ == "__main__":
    sys.exit(selftest() if "--selftest" in sys.argv else selftest())
